File: solutions/go/parsing-log-files/1/parsing_log_files.go
package parsinglogfiles

import (
    "regexp"
)

func IsValidLine(text string) bool {
	re := regexp.MustCompile(`^\[(TRC|DBG|INF|WRN|ERR|FTL)\]`)
    return re.MatchString(text)
}

func SplitLogLine(text string) []string {
	var separatorRegexp = regexp.MustCompile(`<[\~\*\=\-]*>`)
    return separatorRegexp.Split(text, -1)
}

func CountQuotedPasswords(lines []string) int {
	re := regexp.MustCompile(`(?i)"[^"]*password[^"]*"`)
    counter := 0
    for _, line := range lines {
        if re.MatchString(line) {
            counter++
        }
    }
    return counter
}

func RemoveEndOfLineText(text string) string {
	re := regexp.MustCompile(`end-of-line[0-9]+`)
    s := re.ReplaceAllString(text, "")
    return s
}

func TagWithUserName(lines []string) []string {
	re := regexp.MustCompile(`User\s+(\S+)`)
    for i, line := range lines {
        matches := re.FindStringSubmatch(line)
        if len(matches) >= 2 {
            username := matches[1]
            lines[i] = "[USR] " + username + " " + line
        }
    }
    return lines
}

File: solutions/csharp/parsing-log-files/1/ParsingLogFiles.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

public class LogParser
{
    private static readonly Regex SeparatorRegex = new Regex(@"<[\^\*\=\-]*>", RegexOptions.Compiled);
    
    public bool IsValidLine(string text) =>
        Regex.IsMatch(text, @"^\[(TRC|DBG|INF|WRN|ERR|FTL)\]");

    public string[] SplitLogLine(string text) =>
        SeparatorRegex.Split(text);

    public int CountQuotedPasswords(string lines)
    {
        int counter = 0;
        var pattern = @"\""[^\""]*password[^\""]*\""";
        var regex = new Regex(pattern, RegexOptions.IgnoreCase);
        var lineArray = lines.Split(new[] { "\r\n", "\n" }, StringSplitOptions.None);
    
        foreach (var line in lineArray)
        {
            if (regex.IsMatch(line))
            {
                counter++;
            }
        }
        return counter;
    }

    public string RemoveEndOfLineText(string line) =>
        Regex.Replace(line, @"end-of-line[0-9]+", "");

    public string[] ListLinesWithPasswords(string[] lines)
    {
        var pattern = @"\b(password\w+)";
        var regex = new Regex(pattern, RegexOptions.IgnoreCase);

        var result = new List<string>();

        foreach (var line in lines)
        {
            var match = regex.Match(line);
            if (match.Success)
            {
                result.Add($"{match.Groups[1].Value}: {line}");
            }
            else
            {
                result.Add($"--------: {line}");
            }
        }
        return result.ToArray();
    }
}
